README.md
# Docker-Reference

A small package to parse and work with docker images references.

## Getting started

To use docker-reference you should start with creating a `Reference` instance:
``` js
const { Reference } = require('@swimlane/docker-reference');

const imageReference = new Reference({
    domain: 'r.cfcr.io',
    repository: 'codefresh/demochat',
    tag: 'master',
    digest: 'sha256:58fe87ff24d5a3ac35c887dcef82eb619565987c1083282f876c7c2657a5f94e'
});

console.log(imageReference.repositoryUrl); // r.cfcr.io/codefresh/demochat
console.log(imageReference.toString()); // r.cfcr.io/codefresh/demochat:master@sha256:58fe87ff24d5a3ac35c887dcef82eb619565987c1083282f876c7c2657a5f94e
```

## Using parsers

You can also use parser to create your `Reference` instance:
``` js
const { parseQualifiedName } = require('@swimlane/docker-reference');

const imageReference = parseQualifiedName('r.cfcr.io/codefresh/demochat:master@sha256:58fe87ff24d5a3ac35c887dcef82eb619565987c1083282f876c7c2657a5f94e');

console.log(imageReference.domain);     // r.cfcr.io
console.log(imageReference.repository); // codefresh/demochat
console.log(imageReference.tag);        // master
console.log(imageReference.digest);     // sha256:58fe87ff24d5a3ac35c887dcef82eb619565987c1083282f876c7c2657a5f94e
```

There are 3 parsers in the package:
 1. `parseQualifiedName`:

    This should be use when parsing a string representing a qualified image
    reference, meaning the string should contain at least the domain and the
    repository of the image.

 2. `parseFamiliarName`:

    This should be use when parsing a string representing a familiar image
    reference, meaning the string might not contains the domain part and this
    case the domain part would become `docker.io`. Also when the domain is
    missing and the repository contains only one part the repository would be
    prefixed with `library/`.

    **This parser would parse all qualified image reference the same as
    `parseQualifiedName` parser**

3. `parseAll`:

    This should be use when parsing a string representing a familiar image
    reference or identifier.

    **This parser would parse all qualified image reference the same as
    `parseQualifiedName` parser and parse every familiar image reference the
    same as `parseFamiliarName`**

## Grammer for the package

|  | pattaren |
| --- | --- |
| `reference` | `name [ ":" tag ] [ "@" digest ]` |
| `name` | `[domain '/'] path-component ['/' path-component]*` |
| `domain` | `domain-component ['.' domain-component]* [':' port-number]` |
| `domain-component` | `/([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/` |
| `port-number` | `/[0-9]+/` |
| `path-component` | `alpha-numeric [separator alpha-numeric]*` |
| `alpha-numeric` | `/[a-z0-9]+/` |
| `separator` | `/[_.]|__|[-]*/` |
| `tag` | `/[\w][\w.-]{0,127}/` |
| `digest` | `digest-algorithm ":" digest-hex` |
| `digest-algorithm` | `digest-algorithm-component [ digest-algorithm-separator digest-algorithm-component ]*` |
| `digest-algorithm-separator` | `/[+.-_]/` |
| `digest-algorithm-component` | `/[A-Za-z][A-Za-z0-9]*/` |
| `digest-hex` | `/[0-9a-fA-F]{32,}/` (At least 128 bit digest value) |
| `identifier` | `/[a-f0-9]{64}/` |
| `short-identifier` | `/[a-f0-9]{6,64}/` |


CHANGELOG.md
# Changelog

## [2.0.0](https://github.com/swimlane/node-docker-reference/compare/v1.0.0...v2.0.0) (2022-11-01)


### ⚠ BREAKING CHANGES

* removed RE2

### Features

* refactored to library to TypeScript and removed RE2 ([18406e5](https://github.com/swimlane/node-docker-reference/commit/18406e517af75459e16c9daa19a3f40746578fb7))

## 1.0.0 (2022-10-20)


### Bug Fixes

* removed console.logs ([792b619](https://github.com/swimlane/node-docker-reference/commit/792b61922c22e5416e4fa02edb1b888ee26e261a))


tests/digest.spec.js
'use strict';

/* eslint-env mocha */

const chai = require('chai');
const { validateDigest } = require('../lib/digest');

const expect = chai.expect;

describe('Digest Validation', () => {

    it('should validate valid digest with sha256', () => {
        validateDigest('sha256:e58fcf7418d4390dec8e8fb69d88c06ec07039d651fedd3aa72af9972e7d046b');
    });

    it('should validate valid digest with sha384', () => {
        validateDigest('sha384:d3fc7881460b7e22e3d172954463dddd7866d17597e7' +
                       '248453c48b3e9d26d9596bf9c4a9cf8072c9d5bad76e19af801d');
    });

    it('should throw error on empty hex digest', () => {
        expect(() => validateDigest('sha256:')).to.throw('invalid digest format');
    });

    it('should throw error on empty digest', () => {
        expect(() => validateDigest('')).to.throw('invalid digest format');
    });

    it('should throw error on digest with no algorithm', () => {
        expect(() => validateDigest('d41d8cd98f00b204e9800998ecf8427e'))
            .to.throw('invalid digest format');
    });

    it('should throw error on digest with invalid hex', () => {
        expect(() => validateDigest('sha256:d41d8cd98f00b204e9800m98ecf8427e'))
            .to.throw('invalid digest format');
    });

    it('should throw error on digest with shorted hex', () => {
        expect(() => validateDigest('sha256:abcdef0123456789'))
            .to.throw('invalid digest format');
    });

    it('should throw error on digest with shorted hex in other algorithm', () => {
        expect(() => validateDigest('sha512:abcdef0123456789abcdef012345' +
                                    '6789abcdef0123456789abcdef0123456789'))
            .to.throw('invalid checksum digest length');
    });

    it('should throw error on digest with unknown algorithm', () => {
        expect(() => validateDigest('foo:d41d8cd98f00b204e9800998ecf8427e'))
            .to.throw('unsupported digest algorithm');
    });

    it('should throw error on digest with repeated separator', () => {
        expect(() => validateDigest('sha384__foo+bar:d3fc7881460b7e22e3d172954463dddd7866d175' +
                                    '97e7248453c48b3e9d26d9596bf9c4a9cf8072c9d5bad76e19af801d'))
            .to.throw('invalid digest format');
    });
});



tests/digest.spec.ts
/* eslint-env mocha */
import {validateDigest} from '../src/digest';

describe('Digest Validation', () => {

    it('should validate valid digest with sha256', () => {
        validateDigest('sha256:e58fcf7418d4390dec8e8fb69d88c06ec07039d651fedd3aa72af9972e7d046b');
    });

    it('should validate valid digest with sha384', () => {
        validateDigest('sha384:d3fc7881460b7e22e3d172954463dddd7866d17597e7' +
                       '248453c48b3e9d26d9596bf9c4a9cf8072c9d5bad76e19af801d');
    });

    it('should throw error on empty hex digest', () => {
        expect(() => validateDigest('sha256:')).toThrow('invalid digest format');
    });

    it('should throw error on empty digest', () => {
        expect(() => validateDigest('')).toThrow('invalid digest format');
    });

    it('should throw error on digest with no algorithm', () => {
        expect(() => validateDigest('d41d8cd98f00b204e9800998ecf8427e'))
            .toThrow('invalid digest format');
    });

    it('should throw error on digest with invalid hex', () => {
        expect(() => validateDigest('sha256:d41d8cd98f00b204e9800m98ecf8427e'))
            .toThrow('invalid digest format');
    });

    it('should throw error on digest with shorted hex', () => {
        expect(() => validateDigest('sha256:abcdef0123456789'))
            .toThrow('invalid digest format');
    });

    it('should throw error on digest with shorted hex in other algorithm', () => {
        expect(() => validateDigest('sha512:abcdef0123456789abcdef012345' +
                                    '6789abcdef0123456789abcdef0123456789'))
            .toThrow('invalid checksum digest length');
    });

    it('should throw error on digest with unknown algorithm', () => {
        expect(() => validateDigest('foo:d41d8cd98f00b204e9800998ecf8427e'))
            .toThrow('unsupported digest algorithm');
    });

    it('should throw error on digest with repeated separator', () => {
        expect(() => validateDigest('sha384__foo+bar:d3fc7881460b7e22e3d172954463dddd7866d175' +
                                    '97e7248453c48b3e9d26d9596bf9c4a9cf8072c9d5bad76e19af801d'))
            .toThrow('invalid digest format');
    });
});



tests/familiar-name-parser.js
'use strict';

/* eslint-env mocha */

const chai         = require('chai');
const { parseFamiliarName } = require('../lib/parsers');

const expect = chai.expect;

describe('Familiar name parser', () => {

    describe('should parse valid names without errors:', () => {

        it(`Valid name #1`, () => {
            parseFamiliarName('docker/docker');
        });

        it(`Valid name #2`, () => {
            parseFamiliarName('library/debian');
        });

        it(`Valid name #3`, () => {
            parseFamiliarName('debian');
        });

        it(`Valid name #4`, () => {
            parseFamiliarName('docker.io/docker/docker');
        });

        it(`Valid name #5`, () => {
            parseFamiliarName('docker.io/library/debian');
        });

        it(`Valid name #6`, () => {
            parseFamiliarName('docker.io/debian');
        });

        it(`Valid name #7`, () => {
            parseFamiliarName('index.docker.io/docker/docker');
        });

        it(`Valid name #8`, () => {
            parseFamiliarName('index.docker.io/library/debian');
        });

        it(`Valid name #9`, () => {
            parseFamiliarName('index.docker.io/debian');
        });

        it(`Valid name #10`, () => {
            parseFamiliarName('127.0.0.1:5000/docker/docker');
        });

        it(`Valid name #11`, () => {
            parseFamiliarName('127.0.0.1:5000/library/debian');
        });

        it(`Valid name #12`, () => {
            parseFamiliarName('127.0.0.1:5000/debian');
        });

        it(`Valid name #13`, () => {
            parseFamiliarName('thisisthesongthatneverendsitgoesonandonandonthisisthesongthatnev');
        });

        it(`Valid name #14`, () => {
            parseFamiliarName('docker.io/1a3f5e7d9c1b3a5f7e9d1c3b5a7f9e1d3c5b7a9f1e3d5d7c9b1a3f5e7d9c1b3a');
        });
    });

    describe('should parse invalid names without errors:', () => {

        it(`Invalid name #1`, () => {
            expect(() => parseFamiliarName('https://github.com/docker/docker')).to.throw();
        });

        it(`Invalid name #2`, () => {
            expect(() => parseFamiliarName('docker/Docker')).to.throw();
        });

        it(`Invalid name #3`, () => {
            expect(() => parseFamiliarName('-docker')).to.throw();
        });

        it(`Invalid name #4`, () => {
            expect(() => parseFamiliarName('-docker/docker')).to.throw();
        });

        it(`Invalid name #5`, () => {
            expect(() => parseFamiliarName('-docker.io/docker/docker')).to.throw();
        });

        it(`Invalid name #6`, () => {
            expect(() => parseFamiliarName('docker///docker')).to.throw();
        });

        it(`Invalid name #7`, () => {
            expect(() => parseFamiliarName('docker.io/docker/Docker')).to.throw();
        });

        it(`Invalid name #8`, () => {
            expect(() => parseFamiliarName('docker.io/docker///docker')).to.throw();
        });

        it(`Invalid name #9`, () => {
            expect(() => parseFamiliarName(
                '1a3f5e7d9c1b3a5f7e9d1c3b5a7f9e1d3c5b7a9f1e3d5d7c9b1a3f5e7d9c1b3a')).to.throw();
        });

    });
});



src/regexp.ts
/**
 * expression defines a full expression, where each regular expression must follow the previous.
 */
const expression = (...regexps: RegExp[]) => new RegExp(regexps
    .map(re => re.source)
    .join(''));

/**
 * group wraps the regexp in a non-capturing group.
 */
const group = (...regexps: RegExp[]) => new RegExp(`(?:${expression(...regexps).source})`);

/**
 * repeated wraps the regexp in a non-capturing group to get one or more matches.
 */
const optional = (...regexps: RegExp[]) => new RegExp(`${group(...regexps).source}?`);

/**
 * repeated wraps the regexp in a non-capturing group to get one or more matches.
 */
const repeated = (...regexps: RegExp[]) => new RegExp(`${group(...regexps).source}+`);

/**
 * anchored anchors the regular expression by adding start and end delimiters.
 */
const anchored = (...regexps: RegExp[]) => new RegExp(`^${expression(...regexps).source}$`);

/**
 * capture wraps the expression in a capturing group.
 */
const capture = (...regexps: RegExp[]) => new RegExp(`(${expression(...regexps).source})`);

const alphaNumericRegexp = /[a-z0-9]+/;
const separatorRegexp    = /(?:[._]|__|[-]*)/;

const nameComponentRegexp = expression(
    alphaNumericRegexp,
    optional(repeated(separatorRegexp, alphaNumericRegexp))
);

const domainComponentRegexp = /(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])/;

const domainRegexp = expression(
    domainComponentRegexp,
    optional(repeated(/\./, domainComponentRegexp)),
    optional(/:/, /[0-9]+/)
);

const tagRegexp = /[\w][\w.-]{0,127}/;

const digestRegexp = /[a-zA-Z][a-zA-Z0-9]*(?:[-_+.][a-zA-Z][a-zA-Z0-9]*)*[:][a-fA-F0-9]{32,}/;

const anchoredDigestRegexp = anchored(digestRegexp);

const nameRegexp = expression(
    optional(domainRegexp, /\//),
    nameComponentRegexp,
    optional(repeated(/\//, nameComponentRegexp))
);

const anchoredNameRegexp = anchored(
    optional(capture(domainRegexp), /\//),
    capture(nameComponentRegexp, optional(repeated(/\//, nameComponentRegexp)))
);

const referenceRegexp = anchored(
    capture(nameRegexp),
    optional(/:/, capture(tagRegexp)),
    optional(/@/, capture(digestRegexp))
);

const identifierRegexp = /[a-f0-9]{64}/;

const anchoredIdentifierRegexp = anchored(identifierRegexp);
export {
    referenceRegexp,
    anchoredNameRegexp,
    anchoredDigestRegexp,
    anchoredIdentifierRegexp
}


lib/parsers.js
"use strict";

let RE2;
try {
  RE2 = require("re2");
  // Test if native is working
  new RE2(".*").exec("test");  
} catch (err) {  
  RE2 = RegExp;
}

const {
  referenceRegexp,
  anchoredNameRegexp,
  anchoredIdentifierRegexp,
} = require("./regexp");
const { validateDigest, isDigest } = require("./digest");
const { Reference } = require("./reference");

const NAME_MAX_LENGTH = 255;

class InvalidReferenceFormatError extends Error {
  constructor() {
    super("invalid reference format");
    this.name = "InvalidReferenceFormatError";
  }
}

class NameContainsUppercaseError extends Error {
  constructor() {
    super("repository name must be lowercase");
    this.name = "NameContainsUppercaseError";
  }
}

class EmptyNameError extends Error {
  constructor() {
    super("repository name must have at least one component");
    this.name = "EmptyNameError";
  }
}

class NameTooLongError extends Error {
  constructor() {
    super(
      `repository name must not be more than ${NAME_MAX_LENGTH} characters`
    );
    this.name = "NameTooLongError";
  }
}

const DEFAULT_DOMAIN = "docker.io";
const LEGACY_DEFAULT_DOMAIN = "index.docker.io";
const OFFICIAL_REPOSITORY_NAME = "library";

function _parseQualifiedName(regexp, name) {
  const matches = regexp.exec(name);

  if (!matches) {
    if (name === "") {
      throw new EmptyNameError();
    }

    if (regexp.test(name.toLowerCase())) {
      throw new NameContainsUppercaseError();
    }

    throw new InvalidReferenceFormatError();
  }

  if (matches[1].length > NAME_MAX_LENGTH) {
    throw new NameTooLongError();
  }

  let reference;

  const nameMatch = anchoredNameRegexp.exec(matches[1]);
  if (nameMatch && nameMatch.length === 3) {
    reference = {
      domain: nameMatch[1],
      repository: nameMatch[2],
    };
  } else {
    reference = {
      domain: "",
      repository: matches[1],
    };
  }

  reference.tag = matches[2];

  if (matches[3]) {
    validateDigest(matches[3]);
    reference.digest = matches[3];
  }

  return new Reference(reference);
}

exports.parseQualifiedNameOptimized = (name) => {
  return _parseQualifiedName(new RE2(referenceRegexp), name);
};

exports.parseQualifiedName = (name) => {
  return _parseQualifiedName(referenceRegexp, name);
};

function splitDockerDomain(name) {
  let domain;
  let reminder;

  const indexOfSlash = name.indexOf("/");
  if (
    indexOfSlash === -1 ||
    !(
      name.lastIndexOf(".", indexOfSlash) !== -1 ||
      name.lastIndexOf(":", indexOfSlash) !== -1 ||
      name.startsWith("localhost/")
    )
  ) {
    domain = DEFAULT_DOMAIN;
    reminder = name;
  } else {
    domain = name.substring(0, indexOfSlash);
    reminder = name.substring(indexOfSlash + 1);
  }

  if (domain === LEGACY_DEFAULT_DOMAIN) {
    domain = DEFAULT_DOMAIN;
  }

  if (domain === DEFAULT_DOMAIN && !reminder.includes("/")) {
    reminder = `${OFFICIAL_REPOSITORY_NAME}/${reminder}`;
  }

  return [domain, reminder];
}

exports.parseFamiliarName = (name, parseQualifiedNameFunc) => {
  if (anchoredIdentifierRegexp.test(name)) {
    throw new TypeError(
      `invalid repository name (${name}),` +
        `cannot specify 64-byte hexadecimal strings`
    );
  }

  const [domain, remainder] = splitDockerDomain(name);

  let remoteName;
  const tagSeparatorIndex = remainder.indexOf(":");
  if (tagSeparatorIndex > -1) {
    remoteName = remainder.substring(0, tagSeparatorIndex);
  } else {
    remoteName = remainder;
  }

  if (remoteName.toLowerCase() !== remoteName) {
    throw new TypeError(
      `invalid reference format: repository name must be lowercase`
    );
  }

  if (parseQualifiedNameFunc) {
    // ability to define custom park func , we have performance
    // issue with old one and we need a way to override
    // it for prevent use js native regexp
    return parseQualifiedNameFunc(`${domain}/${remainder}`);
  }

  return exports.parseQualifiedName(`${domain}/${remainder}`);
};

exports.parseAll = (name) => {
  if (anchoredIdentifierRegexp.test(name)) {
    return new Reference({ digest: `sha256:${name}` });
  }

  if (isDigest(name)) {
    return new Reference({ digest: name });
  }

  return exports.parseFamiliarName(name);
};


src/parsers.ts
import {anchoredIdentifierRegexp, anchoredNameRegexp, referenceRegexp} from './regexp';
import {isDigest, validateDigest} from './digest';
import {Reference, ReferenceOptions, ReferenceType} from './reference';


const NAME_MAX_LENGTH = 255;

class InvalidReferenceFormatError extends Error {
  constructor() {
    super("invalid reference format");
    this.name = "InvalidReferenceFormatError";
  }
}

class NameContainsUppercaseError extends Error {
  constructor() {
    super("repository name must be lowercase");
    this.name = "NameContainsUppercaseError";
  }
}

class EmptyNameError extends Error {
  constructor() {
    super("repository name must have at least one component");
    this.name = "EmptyNameError";
  }
}

class NameTooLongError extends Error {
  constructor() {
    super(
      `repository name must not be more than ${NAME_MAX_LENGTH} characters`
    );
    this.name = "NameTooLongError";
  }
}

const DEFAULT_DOMAIN = "docker.io";
const LEGACY_DEFAULT_DOMAIN = "index.docker.io";
const OFFICIAL_REPOSITORY_NAME = "library";

function _parseQualifiedName(regexp: RegExp, name: string) {
  const matches = regexp.exec(name);

  if (!matches) {
    if (name === "") {
      throw new EmptyNameError();
    }

    if (regexp.test(name.toLowerCase())) {
      throw new NameContainsUppercaseError();
    }

    throw new InvalidReferenceFormatError();
  }

  if (matches[1].length > NAME_MAX_LENGTH) {
    throw new NameTooLongError();
  }

  let reference: ReferenceOptions;

  const nameMatch = anchoredNameRegexp.exec(matches[1]);
  if (nameMatch && nameMatch.length === 3) {
    reference = {
      domain: nameMatch[1],
      repository: nameMatch[2],
    };
  } else {
    reference = {
      domain: "",
      repository: matches[1],
    };
  }

  reference.tag = matches[2] as ReferenceType;

  if (matches[3]) {
    validateDigest(matches[3]);
    reference.digest = matches[3];
  }

  return new Reference(reference);
}

// noinspection JSUnusedGlobalSymbols
const parseQualifiedNameOptimized = (name: string) => {
  return _parseQualifiedName(new RegExp(referenceRegexp), name);
};

const parseQualifiedName = (name: string) => {
  return _parseQualifiedName(referenceRegexp, name);
};

function splitDockerDomain(name: string) {
  let domain;
  let reminder;

  const indexOfSlash = name.indexOf("/");
  if (
    indexOfSlash === -1 ||
    !(
      name.lastIndexOf(".", indexOfSlash) !== -1 ||
      name.lastIndexOf(":", indexOfSlash) !== -1 ||
      name.startsWith("localhost/")
    )
  ) {
    domain = DEFAULT_DOMAIN;
    reminder = name;
  } else {
    domain = name.substring(0, indexOfSlash);
    reminder = name.substring(indexOfSlash + 1);
  }

  if (domain === LEGACY_DEFAULT_DOMAIN) {
    domain = DEFAULT_DOMAIN;
  }

  if (domain === DEFAULT_DOMAIN && !reminder.includes("/")) {
    reminder = `${OFFICIAL_REPOSITORY_NAME}/${reminder}`;
  }

  return [domain, reminder];
}

const parseFamiliarName = (name: string, parseQualifiedNameFunc?: (name: string)=> string) => {
  if (anchoredIdentifierRegexp.test(name)) {
    throw new TypeError(
      `invalid repository name (${name}),` +
        `cannot specify 64-byte hexadecimal strings`
    );
  }

  const [domain, remainder] = splitDockerDomain(name);

  let remoteName;
  const tagSeparatorIndex = remainder.indexOf(":");
  if (tagSeparatorIndex > -1) {
    remoteName = remainder.substring(0, tagSeparatorIndex);
  } else {
    remoteName = remainder;
  }

  if (remoteName.toLowerCase() !== remoteName) {
    throw new TypeError(
      `invalid reference format: repository name must be lowercase`
    );
  }

  if (parseQualifiedNameFunc) {
    // ability to define custom park func , we have performance
    // issue with old one, and we need a way to override
    // it for prevent use js native regexp
    return parseQualifiedNameFunc(`${domain}/${remainder}`);
  }

  return parseQualifiedName(`${domain}/${remainder}`);
};

// noinspection JSUnusedGlobalSymbols
const parseAll = (name: string) => {
  if (anchoredIdentifierRegexp.test(name)) {
    return new Reference({ digest: `sha256:${name}` });
  }

  if (isDigest(name)) {
    return new Reference({ digest: name });
  }

  return parseFamiliarName(name);
};


export {
  parseQualifiedNameOptimized,
  parseQualifiedName,
  parseFamiliarName,
  parseAll
}


src/reference.ts
export interface ReferenceOptions {
    digest?: string;
    type?: ReferenceType;
    domain?: string;
    repository?: string;
    tag?: string;
}

const typesTemplates = {
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    'digest': (ref: Reference) => `${ref.digest}`,
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    'canonical': (ref: Reference) => `${ref.repositoryUrl}@${ref.digest}`,
    'repository': (ref: Reference) => `${ref.repositoryUrl}`,
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    'tagged': (ref: Reference) => `${ref.repositoryUrl}:${ref.tag}`,
    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
    'dual': (ref: Reference) => `${ref.repositoryUrl}:${ref.tag}@${ref.digest}`
};

export type ReferenceType = keyof typeof typesTemplates

export class Reference {
    get tag(): string | undefined{
        return this._tag;
    }
    get digest(): string | undefined{
        return this._digest;
    }
    get repository(): string | undefined{
        return this._repository;
    }
    get domain(): string | undefined{
        return this._domain;
    }
    get type(): ReferenceType{
        return this._type
    }
    private readonly _domain: string | undefined ;
    private readonly _repository: string | undefined ;
    private readonly _digest: string | undefined ;
    private readonly _type: ReferenceType;
    private readonly _tag: string | undefined ;

    constructor(options: ReferenceOptions) {
        if (!options.repository && !options.domain) {
            if (options.digest) {
                this._digest = options.digest;
                this._type = 'digest';
            } else {
                throw new TypeError('Empty Reference');
            }
        } else if (!options.tag) {
            this._domain = options.domain ?? '';
            this._repository = options.repository ?? '';
            if (options.digest) {
                this._digest = options.digest;
                this._type = 'canonical';
            } else {
                this._type = 'repository';
            }
        } else if (!options.digest) {
            this._domain= options.domain ?? '';
            this._repository = options.repository ?? '';
            this._tag = options.tag;
            this._type= 'tagged';
        } else {
            this._domain = options.domain ?? '';
            this._repository = options.repository ?? '';
            this._tag = options.tag;
            this._digest = options.digest;
            this._type = 'dual';
        }
    }

    toString() {
        return typesTemplates[this._type](this);
    }

    get repositoryUrl() {
        if(this._domain && this._repository) {
            return `${this._domain}/${this._repository}`;
        }
        return ''
    }
}


src/digest.ts
import {anchoredDigestRegexp} from './regexp';

class InvalidDigestFormatError extends Error {
    constructor() {
        super('invalid digest format');
        this.name = 'InvalidDigestFormatError';
    }
}

class UnsupportedAlgorithmError extends Error {
    constructor() {
        super('unsupported digest algorithm');
        this.name = 'UnsupportedAlgorithmError';
    }
}

class InvalidDigestLengthError extends Error {
    constructor() {
        super('invalid checksum digest length');
        this.name = 'InvalidDigestLengthError';
    }
}

const algorithmsSizes  ={
    sha256: 32,
    sha384: 48,
    sha512: 64,
}

type AlgorithmsSizes = keyof typeof algorithmsSizes
type ErrorHandler = <TError extends Error>(t: new () => TError) => void

function checkDigest(digest: string, handleError: ErrorHandler) {
    const indexOfColon = digest.indexOf(':');
    if (indexOfColon < 0 ||
        indexOfColon + 1 === digest.length ||
        !anchoredDigestRegexp.test(digest)) {
        return handleError(InvalidDigestFormatError);
    }

    const algorithm = digest.substring(0, indexOfColon) as AlgorithmsSizes;

    if (!Object.hasOwnProperty.call(algorithmsSizes, algorithm)) {
        return handleError(UnsupportedAlgorithmError);
    }

    if (algorithmsSizes[algorithm] * 2 !== (digest.length - indexOfColon - 1)) {
        return handleError(InvalidDigestLengthError);
    }

    return true;
}

const validateDigest = (digest: string) => {
    checkDigest(digest, (ErrorType) => {
        // eslint-disable-next-line @typescript-eslint/no-throw-literal
        throw new ErrorType();
    });
};

const isDigest = (digest: string) => {
    return checkDigest(digest, () => false);
};

export {
    validateDigest,
    isDigest
}
